cabal-install/Distribution/Client/CmdClean.hs

```diff
@@ -0,0 +1,46 @@
+module Distribution.Client.CmdClean (cleanCommand, cleanAction) where
+
+import Prelude ()
+import Distribution.Client.Compat.Prelude
+
+import Distribution.Client.DistDirLayout
+import Distribution.Client.ProjectOrchestration
+
+import Distribution.Client.Setup
+          ( GlobalFlags, ConfigFlags, ConfigExFlags, InstallFlags, configVerbosity )
+import qualified Distribution.Client.Setup as Client
+import Distribution.Simple.Setup
+          ( HaddockFlags, fromFlagOrDefault )
+import Distribution.Simple.Command
+          ( CommandUI(..) )
+import Distribution.Simple.Utils
+          ( wrapText )
+import Distribution.Verbosity
+          ( normal )
+
+import System.Directory 
+          ( removeDirectoryRecursive )
+
+cleanCommand :: CommandUI (ConfigFlags, ConfigExFlags, InstallFlags, HaddockFlags)
+cleanCommand = Client.installCommand
+  { commandName = "new-clean"
+  , commandSynopsis = "Clean the package store and remove temporary files."
+  , commandUsage = \pname ->
+    "Usage: " ++ pname ++ " new-clean [FLAGS]\n"
+  , commandDescription  = Just $ \_ -> wrapText $
+    "Removes all temporary files created during the building process "
+ ++ "(.hi, .o, preprocessed sources, etc.) and also empties out the "
+ ++ "nix-style package store.\n\n"
+  , commandNotes = Nothing
+  }
+
+cleanAction :: (ConfigFlags, ConfigExFlags, InstallFlags, HaddockFlags) -> [String] -> GlobalFlags -> IO ()
+cleanAction (configFlags, configExFlags, installFlags, haddockFlags) extraArgs globalFlags = do
+    let cliConfig = commandLineFlagsToProjectConfig globalFlags configFlags configExFlags installFlags haddockFlags
+        verbosity = fromFlagOrDefault normal (configVerbosity configFlags)
+    
+    baseCtx <- establishProjectBaseContext verbosity cliConfig
+
+    let distDir = distDirectory (distDirLayout baseCtx)
+
+    removeDirectoryRecursive distDir
```

Comment:
Ahh I see, so using the ProjectOrchestration it looks like we need all this info just to establish the dist dir layout. I strongly suspect however that we could construct the dist dir layout from a whole lot less, and then we wouldn't need to take all these various sets of flags, and would not need to derive from the install command UI.